File: sol/solution/0001-0100/0015/Solution.kt
class Solution {
    fun threeSum(nums: IntArray): List<List<Int>> {
        nums.sort()
        val res = mutableListOf<List<Int>>()
        for (i in 0 until nums.size - 2) {
            if (i > 0 && nums[i] == nums[i - 1]) continue
            var l = i + 1
            var r = nums.size - 1
            while (l < r) {
                val sum = nums[i] + nums[l] + nums[r]
                when {
                    sum < 0 -> l++
                    sum > 0 -> r--
                    else -> {
                        res.add(listOf(nums[i], nums[l], nums[r]))
                        while (l < r && nums[l] == nums[l + 1]) l++
                        while (l < r && nums[r] == nums[r - 1]) r--
                        l++
                        r--
                    }
                }
            }
        }
        return res
    }
}

File: sol/solution/0001-0100/0015/Solution.scala
object Solution {
    def threeSum(nums: Array[Int]): List[List[Int]] = {
        val sorted = nums.sorted
        var res = List[List[Int]]()
        for (i <- 0 until sorted.length - 2) {
            if (i == 0 || sorted(i) != sorted(i - 1)) {
                var l = i + 1
                var r = sorted.length - 1
                while (l < r) {
                    val sum = sorted(i) + sorted(l) + sorted(r)
                    if (sum < 0) l += 1
                    else if (sum > 0) r -= 1
                    else {
                        res ::= List(sorted(i), sorted(l), sorted(r))
                        while (l < r && sorted(l) == sorted(l + 1)) l += 1
                        while (l < r && sorted(r) == sorted(r - 1)) r -= 1
                        l += 1
                        r -= 1
                    }
                }
            }
        }
        res
    }
}
